lib/kafka-producer.js

```diff
@@ -76,24 +76,35 @@ Producer.prototype.getProducer = Promise.method(function (opts, topts) {
 Producer.prototype._produceWrapper = function (producerInstance, topicName,
   partition, value, key, timestamp, optOpaque) {
 
-  if (!this.sr.valueSchemas[topicName]) {
-    // topic not found in schemas, bail early
+  var sendKey = key;
+  var sendValue = value;
+
+  if (key) {
+    if (!this.sr.keySchemas[topicName]) {
+      log.warn('_produceWrapper() :: Warning, did not find topic on SR for key schema:',
+        topicName);
+
+      sendKey = new Buffer(JSON.stringify(key));
+    } else {
+      var typeKey = this.sr.keySchemas[topicName];
+      var schemaIdKey = this.sr.schemaIds[topicName + '-key'];
+      sendKey = this.serialize(typeKey, schemaIdKey, key);
+    }
+  }
 
-    log.warn('_produceWrapper() :: Warning, did not find topic on SR:',
+  if (!this.sr.valueSchemas[topicName]) {
+    log.warn('_produceWrapper() :: Warning, did not find topic on SR for value schema:',
       topicName);
 
-    var bufVal = new Buffer(JSON.stringify(value));
-    return producerInstance.__kafkaAvro_produce(topicName, partition, bufVal,
-      key, timestamp, optOpaque);
+    sendValue = new Buffer(JSON.stringify(value));
+  } else {
+    var typeValue = this.sr.valueSchemas[topicName];
+    var schemaIdValue = this.sr.schemaIds[topicName + '-value'];
+    sendValue = this.serialize(typeValue, schemaIdValue, value);
```

Comment:
these conditions should be a function to less duplicated code

Reply:
Not sure if you agree with the function names or placement thereof, but let me know what you think of it.

Reply:
it's great, thanks